### ISSUE TITLE
Warning about missing `dyn` in `#[ derive() ]`

### ISSUE THREAD

**Issue by najamelan** [rust-lang/rust#52135]
```
warning: trait objects without an explicit `dyn` are deprecated>                                                                      
  -->
26 | #[ derive( Copy, Clone, Eq, PartialEq, Debug, Fail ) ]
   |                                               ^^^^ help: use `dyn`: `dyn (Fail)`
note: lint level defined here
  --> ...path
   |
1  | #![ warn   ( rust_2018_idioms  ) ] // Opt in to warnings about new 2018 idioms
   |              ^^^^^^^^^^^^^^^^
   = note: #[warn(bare_trait_objects)] implied by #[warn(rust_2018_idioms)]
```

This looks wrong to me, and the suggestion doesn't even compile either. Does the warning check whether we are in an `attribute`?

**Comment by Benjamin-L** [comment]
I'm also experiencing this, and it's also worth noting that adding `#[allow(bare_trait_objects)]` to the struct in question doesn't remove the warning.

**Comment by oli-obk** [comment]
I guess derive macro expansions should have the edition of the derive macro crate and not of the one using the `derive` attribute?

**Comment by najamelan** [comment]
@withoutboats  ah, so this would be some code in the crate `failure_derive` causing this warning and the warning would be correct?

`#![ allow( bare_trait_objects ) ]` on top of the file silences the warning.

These are the macro expansions from `failure_derive`. One of them seems to have `dyn` but the other doesn't:

```rust
fn cause(&self) -> ::std::option::Option<&::failure::Fail> {
    match *self {
        ErrorKind::NotAFile => { return None }
        ErrorKind::UnicodeError => { return None }
    }
    None
}

impl Fail for MyError {
   fn cause(&self) -> Option<&dyn Fail> { self.inner.cause() }
   fn backtrace(&self) -> Option<&Backtrace> {
       self.inner.backtrace()
   }
}
```

**Comment by alexcrichton** [comment]
While during triage we placed this on the "Rust 2018 Preview 2" milestone I realize now this is related to the `rust_2018_idioms` lint which is less critical than the compatibility lint, so I'm going to reclassify as the release candidate
